### DISCUSSION TITLE
중간발표 MariaDB

### DISCUSSION THREAD

**Discussion by GeonSoon1** [WHATDO-TEAM/whatdo#59] [발표 자료 관련]
># MariaDB
현재 1번 DB Replication 작업은 완료했고 현재 2번 MHA 구축중입니다.
PPT자료 [Maria DB.pptx](https://github.com/user-attachments/files/17630035/Maria.DB.pptx)

### 1. DB Replication
- 간단한 설명
![image](https://github.com/user-attachments/assets/407aa485-2cf0-480a-84bb-d24ffaa822d6)

&nbsp;

- 자세한 설명
![image](https://github.com/user-attachments/assets/41b7a8d4-eb04-4e9a-b20b-753a47d58849)

이 그림은 **DB 복제(DB Replication)**의 구조를 설명하고 있어요. 그림은 MySQL을 사용하는 MasterDB와 ReplicaDB 사이의 데이터 복제 과정을 보여줍니다.

### MasterDB 측

- **트랜잭션 처리 Thread**: 데이터베이스에서 발생하는 데이터 변경 이벤트(삽입, 삭제, 업데이트 등)를 처리합니다.
- **Binary Log Dump Thread**: Binary Log에서 변경 내용을 읽어와서 Replica Server로 전송하는 역할을 합니다.
- **Binary Log**: MasterDB의 데이터 변경 사항을 기록하는 로그 파일입니다.

### ReplicaDB 측

- **I/O Receiver Thread**: MasterDB에서 전송된 Binary Log 데이터를 받아와서 Relay Log에 기록합니다.
- **Relay Log**: ReplicaDB에서 받은 변경 사항을 기록하는 임시 로그 파일입니다.
- **SQL Applier Thread**: Relay Log에 저장된 변경 사항을 읽어 실제 데이터 파일에 적용합니다.
- **Data File**: 변경 사항이 반영된 최종 데이터가 저장되는 파일입니다.

요약하자면, MasterDB에서 데이터 변경 사항이 발생하면 Binary Log에 기록되고, 이 로그가 ReplicaDB로 전송됩니다. ReplicaDB는 이 로그를 Relay Log에 기록한 뒤, SQL Applier Thread가 해당 변경 사항을 데이터 파일에 적용하여 데이터가 동기화됩니다.
&nbsp;
&nbsp;

### 2. MHA
MHA란 Master 를 대상으로 health check를 통해 상태를 확인하고, 장애 발생시 자동으로 Replica 서버를 Master 서버로 승격시켜준다.
![image](https://github.com/user-attachments/assets/2c12c2ac-5d14-4891-8c6b-31dc4af7cf25)
&nbsp;
![image](https://github.com/user-attachments/assets/9ef6adb1-68ae-4236-b608-6996be116dec)
&nbsp;

**Comment by Kdreamtomaster** [comment]
현재 서버 1개로 구성되어있는데 서버 3개로 각각 DB 1개 씩 띄우고 HAPROXY 나 로드밸런서로 한 아이피 및 포트로 접근 하는것 고려 필요해보입니다.
